Add a local speechbot service

emilyxzhou opened the pull request:
The Rasa bot currently has 2 default assistants to choose from, both defined in `harmoni_bot/src`: rasa_example and rasa_greeter. The assistant can be changed in the configuration file by setting the rasa_assistant param value to the name of the Rasa bot's directory; `rasa_assistant: "rasa_greeter"` would direct the start_rasa_server script to run the server from the `rasa_bot/src/rasa_greeter` directory.

Currently, unit tests and rostests still need to be improved.

emilyxzhou committed:
Set up Rasa service

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Set up Rasa service with temporary data

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Add bot src files to path

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Test  training and running Rasa from Python script

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Remove Rasa server script from launch file

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Create separate script to launch Rasa server

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Add rasa_greeter workspace

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Add dependencies

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Add tests

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Launch rasa server with bash script

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Move script to start rasa server to /scripts

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Merge branch 'develop' of https://github.com/interaction-lab/HARMONI into feature/local-speechbot

emilyxzhou committed:
Add test

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Update rostests

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Merge branch 'develop' of https://github.com/interaction-lab/HARMONI into feature/local-speechbot

# Conflicts:
#	dockerfiles/harmoni/kinetic/base/dockerfile
#	dockerfiles/harmoni/noetic/base/dockerfile

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Update README

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Update requirements.txt

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Set harmoniResponse message value

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Update rostests

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou commented:
Both unit tests and rostests have been updated now, and the Rasa service is fully functional.

micolspitale93 commented on harmoni_dialogues/harmoni_bot/requirements.txt:
```diff
@@ -0,0 +1 @@
+rasa==2.7.1
```
Why the Rasa version differs between the requirements (2.7.1) file and the docker images (2.4.2)?

emilyxzhou replied:
Sorry, the dockerfiles have been updated to 2.7.1 now!

micolspitale93 replied:
Sounds good

micolspitale93 reviewed (commented):
I have tested the rasa_greeter and it works fine for me.
I think that a brief tutorial on how to create a new rasa agent in the harmoni_bot package could be very helpful.

emilyxzhou committed:
Update rasa version

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Set up chatbot pattern example

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Use loop in chatbot example

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Merge branch 'develop' of https://github.com/interaction-lab/HARMONI into feature/local-speechbot

emilyxzhou committed:
Merge branch 'develop' of https://github.com/interaction-lab/HARMONI into feature/local-speechbot

# Conflicts:
#	dockerfiles/harmoni/kinetic/base/dockerfile
#	dockerfiles/harmoni/noetic/base/dockerfile

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Set up voice-based speechbot example

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

RMichaelSwan commented on harmoni_dialogues/harmoni_bot/setup.py:
```diff
@@ -1,12 +1,16 @@
-    package_dir={'': 'src'},
+    package_dir={
+        '': 'src',
+        'harmoni_bot': os.path.join('src', 'harmoni_bot')
```
As far as I know this line isn't needed. harmoni_common_lib doesn't have it and is imported the same way.

RMichaelSwan commented on harmoni_dialogues/harmoni_bot/src/rasa_example/addons/custom_channel.py:
```diff
@@ -0,0 +1,57 @@
+import asyncio
+import inspect
+from sanic import Sanic, Blueprint, response
```
Is this in our dependencies somewhere or is it installed with some other dependency?

emilyxzhou replied:
All of the Rasa dependencies are installed with `pip install rasa`.

RMichaelSwan commented on harmoni_dialogues/harmoni_bot/test/rostest_rasa.py:
```diff
@@ -0,0 +1,98 @@
+from collections import deque
+
+# Specific Imports
+import mock
```
unused?

RMichaelSwan reviewed (commented):
Great work! Once comments are addressed I'll approve it.

emilyxzhou committed:
Remove extra lines

Signed-off-by: emilyxzhou <zhouxemily@gmail.com>

emilyxzhou commented on harmoni_dialogues/harmoni_bot/scripts/start_rasa_server.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+RASA_ASSISTANT=$(rosparam get /rasa/default_param/rasa_assistant)
+RASA_DIR=$(rosparam get /harmoni_bot_dir)
+
+if [ "$RASA_ASSISTANT" == "rasa_example" ] || [ "$RASA_ASSISTANT" == "rasa_greeter" ]
+  then
+    cd "$RASA_DIR"/src/"$RASA_ASSISTANT"
+    rasa train && rasa run
+else
+  echo "Not a valid Rasa bot"
+fi
```
It's easiest to train and run the Rasa models using the Rasa CLI, and I'm not sure if the CLI commands be included directly in roslaunch - that's why I have it in a bash script.

chrismbirmingham commented on harmoni_dialogues/harmoni_bot/scripts/start_rasa_server.sh:
```diff
@@ -0,0 +1,12 @@
+RASA_ASSISTANT=$(rosparam get /rasa/default_param/rasa_assistant)
+RASA_DIR=$(rosparam get /harmoni_bot_dir)
+
+if [ "$RASA_ASSISTANT" == "rasa_example" ] || [ "$RASA_ASSISTANT" == "rasa_greeter" ]
+  then
+    cd "$RASA_DIR"/src/"$RASA_ASSISTANT"
+    rasa train && rasa run
+else
```
This check works for now, but it doesn't work with future bots. Can we dynamically support based on what directories exist in the RASA_DIR?

chrismbirmingham commented on harmoni_dialogues/harmoni_bot/src/rasa_example/actions/actions.py:
```diff
@@ -0,0 +1,27 @@
+#
+#         dispatcher.utter_message(text="Hello World!")
+#
+#         return []
```
Why was this commented out instead of being removed?

chrismbirmingham reviewed (commented):
I'd like to suggest we drop the rasa greeter from being committed here. I think we should include the example bot so there is an example, but that custom bot's should have their own repositories and be imported separately. @RMichaelSwan @micolspitale93 I'd like your opinions on this idea.

emilyxzhou committed:
Move rasa_example to harmoni_models

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Update script to start rasa server

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou committed:
Update README

Signed-off-by: Emily Zhou <zhouxemily@gmail.com>

emilyxzhou commented:
@chrismbirmingham: I've removed `rasa_greeter` and moved `rasa_example` to the models directory. The start_server script has also been updated to dynamically check for existing Rasa workspaces.

chrismbirmingham committed:
Ignore TTS
